packages/lib/src/components/UIElement.tsx

```diff
@@ -101,13 +110,40 @@ export class UIElement<P extends UIElementProps = any> extends BaseElement<P> im
                 break;
             }
 
-            // PM pay button pressed - AnalyticsAction: action: 'log' type:'submit'
-            default: {
+            case ANALYTICS_SUBMIT_STR:
                 // PM pay button pressed - AnalyticsAction: action: 'log' type:'submit'
                 this.props.modules?.analytics.createAnalyticsEvent({
                     event: 'log',
-                    data: { component, type: ANALYTICS_SUBMIT_STR, target: 'payButton', message: 'Shopper clicked pay' }
+                    data: { component, type, target: 'pay_button', message: 'Shopper clicked pay' }
                 });
+                break;
+
+            case ANALYTICS_FOCUS_STR:
+            case ANALYTICS_UNFOCUS_STR:
+                this.props.modules?.analytics.createAnalyticsEvent({
+                    event: 'info',
+                    data: { component, type, target }
+                });
+                break;
+
+            case ANALYTICS_VALIDATION_ERROR_STR: {
+                const { validationErrorCode, validationErrorMessage } = analyticsObj;
+                this.props.modules?.analytics.createAnalyticsEvent({
+                    event: 'info',
+                    data: { component, type, target, validationErrorCode, validationErrorMessage }
+                });
+                break;
+            }
+
+            case ANALYTICS_SELECTED_STR:
+                this.props.modules?.analytics.createAnalyticsEvent({
```

Comment:
Isn't possible to move all this analytics logic to the analytics module?

The UIElement could just pass the required information back to the module, and all the logic about verifying events, and creating them, etc,  would be located there.

That would keep the logic in the UIElement more simple, and most of the analytics code would be in their right place, easy to find and easier to be tested too.

Reply:
I'll give it a try. There will still always be component related logic that will have to be applied (what is the component's `type`, is it a stored Card?) - things a UIElement "knows". So we will always need this `submitAnalytics` function. But, I agree the generation of the analytics objects, based on type, should be part of the Analytics Module